File: src/shared/kartEngine/KartShared/UnityEngine/RaycastHit.lua
-- UnityEngine.RaycastHit结构体的Roblox实现
-- 模拟Unity的RaycastHit结构

local RaycastHit = {}
RaycastHit.__index = RaycastHit

local UnityMath = require(script.Parent.Parent.UnityMath)
local Vector3 = UnityMath.Vector3
local LayerMask = require(script.Parent.LayerMask)

function RaycastHit.new(robloxRaycastResult)
    local self = setmetatable({}, RaycastHit)
    
    if robloxRaycastResult then
        -- 基本属性
        self.distance = robloxRaycastResult.Distance
        self.point = Vector3.new(robloxRaycastResult.Position.X, robloxRaycastResult.Position.Y, robloxRaycastResult.Position.Z)
        self.normal = Vector3.new(robloxRaycastResult.Normal.X, robloxRaycastResult.Normal.Y, robloxRaycastResult.Normal.Z)
        
        -- GameObject相关属性
        local instance = robloxRaycastResult.Instance
        local layer = LayerMask.GetLayer(instance)
        
        -- 模拟Unity的collider结构
        self.collider = {
            gameObject = {
                layer = layer,
                name = instance.Name,
                transform = instance
            }
        }
        
        -- 兼容性属性
        self.transform = instance
        self.rigidbody = instance
        self.layer = layer  -- 方便直接访问layer
    else
        -- 空的RaycastHit
        self.distance = 0
        self.point = Vector3.zero
        self.normal = Vector3.zero
        self.collider = nil
        self.transform = nil
        self.rigidbody = nil
        self.layer = 0
    end
    
    return self
end

return RaycastHit

File: src/shared/kartEngine/KartShared/UnityEngine/RaycastHit.ts
// UnityEngine.RaycastHit结构体的Roblox实现
// 模拟Unity的RaycastHit结构

import { UnityVector3 } from "../UnityMath";
import { LayerMask } from "./LayerMask";

export interface Collider {
    gameObject: {
        layer: number;
        name: string;
        transform: Instance;
    };
}

export class RaycastHit {
    public distance: number;
    public point: UnityVector3; // Vector3 type
    public normal: UnityVector3; // Vector3 type
    public collider: Collider | undefined;
    public transform: Instance | undefined;
    public rigidbody: Instance | undefined;
    public layer: number;

    constructor(robloxRaycastResult?: RaycastResult) {
        if (robloxRaycastResult) {
            // 基本属性
            this.distance = robloxRaycastResult.Distance;
            this.point = new UnityVector3(
                robloxRaycastResult.Position.X,
                robloxRaycastResult.Position.Y,
                robloxRaycastResult.Position.Z
            );
            this.normal = new UnityVector3(
                robloxRaycastResult.Normal.X,
                robloxRaycastResult.Normal.Y,
                robloxRaycastResult.Normal.Z
            );

            // GameObject相关属性
            const instance = robloxRaycastResult.Instance;
            const layer = LayerMask.GetLayer(instance);

            // 模拟Unity的collider结构
            this.collider = {
                gameObject: {
                    layer: layer,
                    name: instance.Name,
                    transform: instance
                }
            };

            // 兼容性属性
            this.transform = instance;
            this.rigidbody = instance;
            this.layer = layer; // 方便直接访问layer
        } else {
            // 空的RaycastHit
            this.distance = 0;
            this.point = UnityVector3.zero;
            this.normal = UnityVector3.zero;
            this.collider = undefined;
            this.transform = undefined;
            this.rigidbody = undefined;
            this.layer = 0;
        }
    }

    public static create(robloxRaycastResult?: RaycastResult): RaycastHit {
        return new RaycastHit(robloxRaycastResult);
    }
}

export default RaycastHit;